src/webview/generateInboundEventEditSequence.js

```diff
@@ -0,0 +1,206 @@
+/* @flow strict-local */
+// $FlowFixMe[untyped-import]
+import isEqual from 'lodash.isequal';
+import invariant from 'invariant';
+
+import type { MessageListElement, Narrow, GetText } from '../types';
+import { ensureUnreachable } from '../generics';
+import { keyFromNarrow } from '../utils/narrow';
+import type { BackgroundData } from './MessageList';
+import messageListElementHtml from './html/messageListElementHtml';
+
+const NODE_ENV = process.env.NODE_ENV;
+
+export type Insert = {|
+  type: 'insert',
+  html: string,
+  index: number,
+|};
+
+export type Delete = {|
+  type: 'delete',
+  index: number,
+|};
+
+export type Replace = {|
+  type: 'replace',
+  html: string,
+  index: number,
+|};
+
+type Edit = Insert | Delete | Replace;
+
+export type EditSequence = $ReadOnlyArray<Edit>;
+
+/**
+ * A compare function to sort a list of MessageListElements by `.key`.
+ *
+ * See the `MessageListElement.key` type definition.
+ */
+function compare(a: MessageListElement, b: MessageListElement) {
+  const firstResult = Math.sign(a.key[0] - b.key[0]);
+  if (firstResult !== 0) {
+    return firstResult;
+  } else {
+    return Math.sign(a.key[1] - b.key[1]);
+  }
+}
+
+function areElementsInOrder(elements: $ReadOnlyArray<MessageListElement>): boolean {
+  return elements.every((elem, i, arr) => i === 0 || compare(arr[i - 1], arr[i]) === -1);
+}
+
+function doElementsDifferInterestingly(
+  oldElement,
+  newElement,
+  oldBackgroundData,
+  newBackgroundData,
+): boolean {
+  if (oldElement.type !== newElement.type) {
+    return true;
+  }
+  switch (oldElement.type) {
+    case 'time':
+      // TODO(?): False positives on `.subsequentMessage.content` changes
+      return !isEqual(oldElement, newElement);
+    case 'header':
+      // TODO(?): False positives on `.subsequentMessage.content` changes
+      return !isEqual(oldElement, newElement);
+    case 'message': {
+      invariant(newElement.type === 'message', 'oldElement.type equals newElement.type');
+
+      // The WebView updates its "read" state through a different event
+      // type, WebViewInboundEventMessagesRead. So, ignore that flag here,
+      // but do notice any other flags changing.
+      // TODO(?): Maybe actually handle 'read' here?
+      const equalFlagsExcludingRead =
+        isEqual(oldElement, newElement)
+        // TODO: Flags are metadata on a message, not "background data".
+        //   Attach them to the MessageListElement object itself, where we
+        //   construct it, then simplify this function. It's surprising to
+        //   see "background data" in this function's inputs.
```

Comment:
This may be a reasonable direction. But I think where it really leads is to keeping the flags directly on the Message objects. That's because their presence on BackgroundData is currently used in a couple more places than just rendering HTML for the message list:
* in handleOutboundEvents for mark-as-read (though only the `read` flag)
* in handleOutboundEvents to pass to the message action sheet, which needs the `starred` flag

For the flags other than `read`, I don't think there's any fundamental reason not to keep them on the message object. It might be an annoying refactor to get there.

For `read`, because that flag changes so often, I think it probably is significant for performance that we avoid having to change all the Message objects when some messages get marked as read. Looking at the flags model now and thinking about it, the right strategy might be to deduplicate that flag with the unreads model which we already have. I've made a note of that, as I have some other unreads-model work on my to-do list which will help enable making that change.

Reply:
I can delete the TODO in the next revision if you'd like. For this one, I'm giving `doElementsDifferInterestingly` a jsdoc with some examples of concerns I was thinking of separating it from. See what you think:

```js
/**
 * Check for differences in data belonging to two MessageListElements
 *
 * Ignores all background data that doesn't belong to the two particular
 * elements and could plausibly affect all elements: language, narrow,
 * dark/light theme, current date, user's role (e.g., admin), etc.
 */
```

Reply:
Sure, that interface makes sense. (And I see you've made this TODO comment a `TODO(?)` in the new revision.)

I'm not sure this will be our final answer to where we put the boundary between background data and message-list elements, and to how we determine when we need to rerender things. But we'll see as we measure performance and as we try things out concretely.